### PR TITLE
Stats: Add Today and Yesterday to date range shortcuts

### PR DESCRIPTION
**Pull request by dognose24**
Related to https://github.com/Automattic/red-team/issues/272

## Proposed Changes

* Add shortcut `Today` and `Yesterday` to the date range picker.

## Why are these changes being made?

* To have the ability to view a single day's stats quickly, we plan to add `Today` and `Yesterday` as an easy way to check recent stats.

## Testing Instructions

* Spin this change up with the Calypso Live branch.
* Navigate to Stats > Traffic page.
* Append the feature flag to the URL: `?flags=stats/new-date-filtering`.
* Click on the date range picker.
* Ensure the dropdown list has `Today` and `Yesterday` options.
* Click on the two new options and ensure the shortcuts work well.

截圖 2024-11-21 凌晨12 00 04

 * Remove the feature flag from the URL.
 * Ensure today and yesterday are displayed as the date on the date control.

截圖 2024-11-21 凌晨12 40 13

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: client/components/date-range/shortcuts.tsx [hunk 1]
```diff
@@ -92,7 +115,10 @@ const DateRangePickerShortcuts = ( {
 		const today = siteToday.clone().startOf( 'day' );
 		const daysInRange = Math.abs( endDate.diff( startDate, 'days' ) );
 		const shortcut = shortcutList.find( ( element ) => {
-			if ( endDate.isSame( today, 'day' ) && daysInRange === element.range ) {
+			if (
+				( endDate.isSame( today, 'day' ) || element.offset === 1 ) &&
```
Comment (dognose24) [inline client/components/date-range/shortcuts.tsx:119]: Temporarily use the `element.offset === 1` check for `yesterday` as there might be a mismatch with the local timezone unless we also offset on the date range picker. cc @kangzj

截圖 2024-11-20 凌晨12 21 01

**Review by kangzj — APPROVED** [review]
I added in some tricks to set `period` to `hour`, when the date range is a single day. It feels like this should be done in the `shortcuts` components, where there are all the dependencies like period, offset and range. However, this is not in the scope of the PR and could be looked after we shipped the full project.

Also I don't see the reason why there are two sets of `shortcutList`, which might need some refactoring as well. Again, not urgent, and could be addressed later.

File: client/components/date-range/shortcuts.tsx [hunk 2]
```diff
@@ -40,6 +41,7 @@ const DateRangePickerShortcuts = ( {
 	const normalizedStartDate = startDate ? normalizeDate( startDate ) : null;
 	const normalizedEndDate = endDate ? normalizeDate( endDate ) : null;
 
+	// TODO: Receive this list from the parent component.
```
Comment (kangzj) [inline client/components/date-range/shortcuts.tsx:44]: I think the full shortcut list could possibly live in the component, and caller could just pass in the list of ids to get them configured.

Comment (dognose24) [reply client/components/date-range/shortcuts.tsx:44]: Both ways are good. We just need to ensure there is only one source of the shortcuts list.

**Pull request merged**